backend/pkg/api/instances.go

```diff
@@ -631,3 +639,108 @@ func (api *API) instanceStatusHistoryQuery(instanceID, appID, groupID string, li
+		t = &now
+	}
+
+	if duration == nil {
+		d := time.Hour
+		duration = &d
+	}
+
+	// Convert duration to string
+	interval := fmt.Sprintf("%d milliseconds", int(duration.Milliseconds()))
+
+	return goqu.From("instance_application").
+		Select(
+			goqu.C("last_check_for_updates").As("timestamp"),
+			goqu.C("c.name").As("channel_name"),
+			goqu.Case().
+				When(goqu.C("c.arch").Eq(1), "AMD64").
+				When(goqu.C("c.arch").Eq(2), "ARM").
+				As("arch"),
+			goqu.C("version"),
+			goqu.L("COUNT(*)").As("instances")).
+		Join(goqu.T("groups"), goqu.On(goqu.C("group_id").Eq(goqu.C("groups.id")))).
+		Join(goqu.T("channel"), goqu.On(goqu.C("groups.channel_id").Eq(goqu.C("channel.id")))).
+		Join(goqu.T("package"), goqu.On(goqu.C("channel.package_id").Eq(goqu.C("package.id")))).
+		Where(
+			goqu.C("last_check_for_updates").Gt(goqu.L("?::timestamp - ?::interval", t.Format(time.RFC3339), interval)),
+			goqu.C("last_check_for_updates").Lte(t.Format(time.RFC3339))).
+		GroupBy("last_check_for_updates", "c.name", "c.arch", "version")
+}
+
+// GetInstanceFacts returns an InstanceFact table with all instances that have
+// been previously been checked in.
+func (api *API) GetInstanceFacts() ([]InstanceFact, error) {
+	query := goqu.From("instance_fact").
+		Select(goqu.L("*"))
+
+	rows, err := api.db.Query(query.ToSQL())
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var instances []InstanceFact
+	for rows.Next() {
+		var instance InstanceFact
+		err = rows.Scan(&instance.Timestamp, &instance.ChannelName, &instance.Arch, &instance.Version, &instance.Instances)
+		if err != nil {
+			return nil, err
+		}
+		instances = append(instances, instance)
+	}
+
+	return instances, nil
+}
+
+// GetInstanceFactsByTimestamp returns an InstanceFact table of instances matching a
+// given timestamp value.
+func (api *API) GetInstanceFactsByTimestamp(t time.Time) ([]InstanceFact, error) {
+	query := goqu.From("instance_fact").
+		Select(goqu.L("*")).
+		Where(
+			goqu.C("timestamp").Eq(t.Format(time.RFC3339)))
+
+	rows, err := api.db.Query(query.ToSQL())
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var instances []InstanceFact
+	for rows.Next() {
+		var instance InstanceFact
+		err = rows.Scan(&instance.Timestamp, &instance.ChannelName, &instance.Arch, &instance.Version, &instance.Instances)
+		if err != nil {
+			return nil, err
+		}
+		instances = append(instances, instance)
+	}
+
+	return instances, nil
+}
+
+// updateInstanceFact updates the instance_fact table with instances checked
+// in during a given duration from a given time.
+func (api *API) updateInstanceFact(t *time.Time, duration *time.Duration) error {
+	insertQuery, _, err := goqu.Insert("instance_fact").
+		Cols("timestamp", "channel_name", "arch", "version", "instances").
+		FromQuery(api.instanceFactQuery(&t, &duration)).
+		ToSQL()
+
+	if err != nil {
+		return err
+	}
+
+	_, err = api.db.Exec(insertQuery)
+	if err != nil {
+		return err
+	}
```

Comment:
```suggestion
	if err != nil {
		return err
	}
	return nil
```
Build failure otherwise because a return is needed

Reply:
Can't believe I missed this thanks!

Reply:
Pointed out by https://github.com/flatcar/nebraska/actions/runs/5416821950/jobs/9847049099?pr=656